### DISCUSSION TITLE
Relaunch daemon after update

### DISCUSSION THREAD

**Discussion by nishant-nf** [sparkle-project/Sparkle#2572] [Q&A]
Our app is also running a daemon. We load a `.plist` file containing
```
<dict>
	<key>KeepAlive</key>
	<true/>
	<key>ProgramArguments</key>
	<array>
		<string>/Applications/NightfallAIAgent.app/Contents/MacOS/NightfallAIAgent</string>
	</array>
	<key>Label</key>
	<string>ai.nightfall.endpoint</string>
	<key>RunAtLoad</key>
	<true/>
</dict>
```
We're trying to update the app in the background. Before the app is updated, we're unloading the job. This works with the following code, but the daemon reload isn't working. How should we be reloading the job after update?
Here's the `updaterDelegate`
```
class UpdaterDelegate: NSObject, SPUUpdaterDelegate {
    func updater(_ updater: SPUUpdater, willInstallUpdateOnQuit item: SUAppcastItem, immediateInstallationBlock immediateInstallHandler: @escaping () -> Void) -> Bool {
        
        manageDaemon(shouldRegister: false, completion: nil)
        return true
    }
    
    func updaterWillRelaunchApplication(_ updater: SPUUpdater) {
        manageDaemon(shouldRegister: true, completion: nil)
    }
    
    func manageDaemon(shouldRegister: Bool, completion: ((Bool) -> Void)?) {
            let plistPath =  "/Library/LaunchDaemons/ai.nightfall.agent.plist"
            let task = Foundation.Process()
            task.launchPath = "/bin/launchctl"

            if shouldRegister {
                task.arguments = ["load", plistPath]
            } else {
                task.arguments = ["unload", plistPath]
            }

            task.terminationHandler = { process in
                completion?(process.terminationStatus == 0)
            }

            task.launch()
        }
}
```

`updaterWillRelaunchApplication` is this the right place to load the `.plist` file again? We're looking to support macOS 12+ so haven't considered using `SMAppService` yet. The update is a zip of the `.app`.
I haven't tried using a `.pkg` which might help on this front since it already includes a script to load the daemon. Would it still require user intervention if running as root or could it be silently installed?

**Comment by zorgiepoo** [comment]
Package installations cannot be done silently and aren't generally recommended.

SMJobBless is available on older OS's where SMAppService isn't (assuming this is a root running daemon, not a user running agent; otherwise there should be other alternatives).

`updaterWillRelaunchApplication` is before the app relaunches from a Sparkle update, not after. Also it's not called if an update is installed on quit but the user does not relaunch the app. There is not a delegate callback for every scenario an app update is installed.

Furthermore, this approach of using Sparkle is wrong. The user can manually update your application without Sparkle. So whatever you need to do needs to generally work in situations where Sparkle isn't used.

**Reply by nishant-nf** [reply]
@zorgiepoo we're looking to update the app in the background without any user intervention. We're fine with not using a package as long as we're able to find a feasible flow using Sparkle. 
What would be the recommended way to restart the app after the update? 
Could you also elaborate on why this approach of using Sparkle is wrong?

**Reply by zorgiepoo** [reply]
I was referring to unloading/loading the launchd job. How do you do that when the user downloads a new version of your app from your website and then replaces the old app with the new one? In other words, how do you handle when the user updates your app _outside of_ Sparkle.

**Reply by zorgiepoo** [reply]
The reason I bring this up is because if you handle that update situation (e.g. by realizing the bundle version increased), then you will realize handling that update does not involve Sparkle. Therefore, you should not integrate Sparkle in your process of unloading/loading a daemon generally.

**Reply by nishant-nf** [reply]
that's currently handled by the post install script in our package.

**Reply by zorgiepoo** [reply]
Yeah typically developers would then serve their packages (pkg) as their updates via Sparkle then. If you want to move off [package installations](https://sparkle-project.org/documentation/package-updates/) (recommended), you should do so everywhere, including website distribution.

Sparkle itself does not install or rely on a true "termination handler" in your app when it is going to be updated, for various reasons. Another thing you probably hadn't considered is if some other updater app reads the Sparkle feed and updates your app.
